src/mpl/src/gpu/mpl_gpu_cuda.c

```diff
@@ -179,12 +135,128 @@ int MPL_gpu_free(void *ptr)
     return MPL_ERR_GPU_INTERNAL;
 }
 
-int MPL_gpu_init()
+int MPL_gpu_init(int local_size, int local_rank, int need_thread_safety)
 {
-    return MPL_SUCCESS;
+    int i, mpl_err = MPL_SUCCESS;
+
+    node_local_size = local_size;
+    self_local_rank = local_rank;
+    cuda_ipc_handle_trees = MPL_malloc(sizeof(MPL_gavl_tree_t) * local_size, MPL_MEM_OTHER);
+    assert(cuda_ipc_handle_trees != NULL);
+    for (i = 0; i < local_size; ++i) {
+        mpl_err =
+            MPL_gavl_tree_create(sizeof(gpu_ipc_handle_obj_s), need_thread_safety,
+                                 &cuda_ipc_handle_trees[i]);
+        if (mpl_err != MPL_SUCCESS) {
+            MPL_gpu_finalize();
+            break;
+        }
+    }
+    return mpl_err;
 }
 
 int MPL_gpu_finalize()
 {
+    if (cuda_ipc_handle_trees) {
+        int i;
+        for (i = 0; i < node_local_size; ++i)
```

Comment:
```suggestion
        for (int i = 0; i < node_local_size; ++i)
```

Reply:
will this cause compilation error when user uses C90 standard?

Reply:
We assume C99 in mpich.

Reply:
got it